proposals/IpfsGatewayHttpApi.md

```diff
@@ -0,0 +1,134 @@
+
+@lidel suggests they could provide a list of issues related to existing API being overly complex, 
+and then js-ipfs-http-client acting as a programmatic interface to cover those shortcomings hurting us even more 
+(breaking API changes, maintenance cost, people confused why they can't just use regular XHR/fetch and need a dedicated client lib).
+
+
+## Project definition
+#### Brief plan of attack
+
+- Design RESTful HTTP semantics on top of content-paths
+- Settle on appropriate serial representation of DAGs (e.g. CAR, DAR, something else).
+- Implement read-only API in HTTP gateway, and deploy it
+- Implement JS client library: request wrappers and IPLD decoding/manipulation
+- Implement publishing API (ie. HTTP POST), but deny from our public gateways
+- Modularise HTTP gateway to permit dapp devs to embed it their back-ends
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+Our gateway serves arbitrary DAGs over simple HTTP requests. 
+Web-apps can fetch, manipulate, and render IPLD data with the standard JS XHR or [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) without the need for any client library. 
+Backends that embed/extend the gateway also support publishing and pinning arbitrary IPLD data.
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+Dapp and other developers are creating clients that interact with IPLD data natively, with most application logic implemented client-side. 
+Our gateway metrics show growth in use of IPLD data. 
+Some dapps are running their own backends embedding a gateway, to support pinning data critical to their application. 
+Infura and others provide hosting services for gateways, as well as nodes.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- Hypotheses about Dapps being an important market, or HTTP being a more accessible API might prove false
+- API design could get stuck in a morass of discussion about standards
+- Hard-core decentralization people object to the imperfect decentralization of using a gateway
+
+#### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+- Just add more RPCs for DAG traversal to go-ipfs, that Infura nodes can then serve. Expose these through JS and other client libraries.
+- Our gateways provide the read-only go-ipfs RPC API to the public, and establish that as standard
+- Amazing technical work so that embedding js-ipfs is both easier and more performant than using an HTTP API
+
+#### Dependencies/prerequisites
+
+- JS-IPLD libraries capable of core encodings and IPLD representation
+- IPLD selectors (for non-trivial DAG traversal)
```

Comment:
just the golang / server side implementation, right?

Reply:
JS implementation will be necessary for sub-DAG queries from the client, which would be needed for complex cases. So no, not just server side. But also we needn't block creating and exposing this API on that JS selector implementation (I gather it doesn't exist?). It will just be more useful once JS clients can express queries.